# **Práctica #1: Modelo para predicción de lluvia**

*Centro Universitario de Ciencias Exactas e Ingenierías*

*División de Tecnologías para la Integración Ciber-Humana*

*Ingeniería Biomédica*

<br>

*Mtra. Sofía Alejandra Aguilar Valdez*

2 de septiembre de 2022

## **Información del equipo**

```NOMBRES:``` \\
Victor Alvarado Aparicio \\
Dayana Analy Pacheco Bañuelos \\
Leonardo Juàrez Zucco

```CÓDIGOS:``` \\
215767685 \\
217535226 \\
220285869

```LINK REPOSITORIO:``` https://github.com/L-Rittmeister/U1_Team.git



## **Contenido**



1.   Resumen
2.   Marco teórico
3.   Objetivos
4.   Materiales y métodos
5.   Resultados
6.   Discusión
7.   Conclusiones
8.   Referencias



# **1. Resumen**
---
En este trabajo se pretendía realizar un modelo de predicción de la lluvia utilizando datos de la estación meteorológica del aeropuerto de Seattle, para este modelo teorizamos utilizar 4 entradas de información para las 4 variables con las que contábamos que eran la probabilidad de precipitación, las temperaturas máxima y mínima y una confirmación si llovió o no con una capa oculta. La metodología usada empieza con la descarga datos, después la construcción del modelo con las entradas y variables mencionadas antes, además declaramos la función de costo nn.CrossEntropyLoss(), y también un algoritmo de optimización optim.Adam(model.parameters(), lr=0.001), para el ajuste de parámetros de la red, por último el entrenamiento del modelo y la evaluación de este. Al concluir la parte de la evaluación y prueba de la red, fue satisfactorio obtener el resultado fue el esperado.



# **2. Marco teórico**
---
El modelo de predicción de la lluvia se basa en la información de la estación meteorológica del aeropuerto de Seattle, Washington, Estados Unidos, la cual se encuentra en el nivel del mar, con una latitud de 47.44° N y una longitud de 122.30° W.
Para poder realizar un modelo funcional de una red neuronal es importante empezar con la descarga de los datos, para este caso se utilizó la librería pandas, la cual nos permite leer los datos de un archivo csv, en este caso se utilizó el archivo weatherHistory.csv. Luego de esto se procedió a la limpieza de los datos, para esto se eliminaron las columnas que no se utilizarían en el modelo, como la fecha, la descripción del clima y la precipitación acumulada. Después de esto se procedió a la construcción del modelo, para esto se utilizó la librería torch, la cual nos permite crear una red neuronal, en este caso se utilizó una red neuronal de 4 capas, la primera capa es la capa de entrada, la segunda capa es la capa oculta, la tercera capa es la capa de salida y la cuarta capa es la capa de activación. Después de esto se procedió a la definición de la función de costo, para esto se utilizó la librería torch.nn, la cual nos permite definir la función de costo, en este caso se utilizó la función de costo nn.CrossEntropyLoss(), la cual nos permite calcular la pérdida de la red neuronal. Después de esto se procedió a la definición del algoritmo de optimización, para esto se utilizó la librería torch.optim, la cual nos permite definir el algoritmo de optimización, en este caso se utilizó el algoritmo de optimización optim.Adam(model.parameters(), lr=0.001), el cual nos permite ajustar los parámetros de la red neuronal. Después de esto se procedió al entrenamiento del modelo, para esto se utilizó la librería torch, la cual nos permite entrenar la red neuronal, en este caso se utilizó el método train(), el cual nos permite entrenar la red neuronal. Después de esto se procedió a la evaluación del modelo, para esto se utilizó la librería torch, la cual nos permite evaluar la red neuronal, en este caso se utilizó el método test(), el cual nos permite evaluar la red neuronal. Al finalizar la evaluación del modelo se procedió a la prueba del modelo, para esto se utilizó la librería torch, la cual nos permite probar la red neuronal, en este caso se utilizó el método predict(). Al finalizar la prueba del modelo se obtuvo un resultado satisfactorio, el cual nos permite concluir que el modelo es funcional.


# **3. Objetivos**
---

### Procesar los datos y crear los objetos e instancias necesarios para poder hacer correr un modelo de predicción basado en una red neuronal que sea capaz de predecir si lloverá o no en Seattle, Washington, USA, utilizando los datos de la estación meteorológica del aeropuerto de Seattle.
### Entrenar el modelo con los datos provistos por la fuente dada y probarlo con los datos de prueba. Evaluar el modelo y considerar un posible replanteamiento si se presentan fenómenos como underfitting u overfitting, así como problemas de compatibilidad de datos, arquitectura del modelo e implementación de las funciones de entrenamiento y validación.


# **4. Materiales y métodos**
---
---
## *Materiales*
---
Describir este conjunto de datos [[1]](https://www.kaggle.com/code/fatmakursun/rain-forecasting-with-artificial-neural-network/data).

## *Métodos*
---
### 4.1. Esquema de metodología
---

In [2]:
#Mounting drive storage to access data (if needed/running on collab then uncomment)
# Mount Google Drive
#from google.colab import drive
#drive.mount('/content/drive')

In [3]:
# Import Libraries
import pandas as pd
import numpy as np
from tqdm.notebook import trange, tqdm
import torch
from torch import nn
from torch.utils.data import Dataset
from torch.utils.data import random_split, DataLoader

In [ ]:
#Pytorch and CUDA installation test
torch.randn(5).cuda()

In [5]:
#Load the dataset on a pandas dataframe
DataF = pd.read_csv("seattleWeather_1948-2017.csv")
DataF

DATE  PRCP  TMAX  TMIN   RAIN
0      1948-01-01  0.47    51    42   True
1      1948-01-02  0.59    45    36   True
2      1948-01-03  0.42    45    35   True
3      1948-01-04  0.31    45    34   True
4      1948-01-05  0.17    45    32   True
...           ...   ...   ...   ...    ...
25546  2017-12-10  0.00    49    34  False
25547  2017-12-11  0.00    49    29  False
25548  2017-12-12  0.00    46    32  False
25549  2017-12-13  0.00    48    34  False
25550  2017-12-14  0.00    50    36  False

[25551 rows x 5 columns]

In [6]:
#Drop NaN value
DataF.dropna(inplace=True)
DataF.reset_index(inplace=True, drop=True)

In [64]:
#eliminate Date column from the dataset
DataF.drop(['DATE'], axis=1, inplace=True)


In [65]:
#Transform the data to float
DataF['PRCP'] = DataF['PRCP'].astype(float)
DataF['TMAX'] = DataF['TMAX'].astype(float)
DataF['TMIN'] = DataF['TMIN'].astype(float)
DataF['RAIN'] = DataF['RAIN'].astype(float)


In [66]:
#Convert the dataframe to a numpy array
DataF = DataF.to_numpy()

In [67]:
#Create the dataset class using the PRCP, TMAX, TMIN as inputs and RAIN as label
class Dataset(Dataset):
    def __init__(self, data):
        self.data = data
        self.len = data.shape[0]
        self.x_data = torch.from_numpy(self.data[:, :3])
        self.y_data = torch.from_numpy(self.data[:, 3])
        self.y_data = self.y_data.type(torch.LongTensor)
    def __getitem__(self, index):
        return self.x_data[index], self.y_data[index]
    def __len__(self):
        return self.len


In [68]:
#Create the train and test dataset using the Dataset class and the data from the dataframe with 90% of the data for training and 10% for testing
train_data = Dataset(DataF[:int(0.9*len(DataF))])
test_data = Dataset(DataF[int(0.9*len(DataF)):])


In [69]:
#Create the train and test dataloader using the train and test dataset
train_loader = DataLoader(dataset=train_data, batch_size=64, shuffle=True)
test_loader = DataLoader(dataset=test_data, batch_size=64, shuffle=True)


In [70]:
#Create the model class using 3 inputs and 2 hidden layers with 3 neurons each and 1 output
class Model(nn.Module):
    def __init__(self, n_inputs):
        super(Model, self).__init__()
        self.hidden1 = nn.Linear(n_inputs, 3)
        self.hidden2 = nn.Linear(3, 3)
        self.hidden3 = nn.Linear(3, 3)
        self.output = nn.Linear(3, 2)
        self.relu = nn.ReLU()
    def forward(self, x):
        x = self.relu(self.hidden1(x))
        x = self.relu(self.hidden2(x))
        x = self.relu(self.hidden3(x))
        x = self.output(x)
        return x


In [71]:
#Create the model using the model class and the number of inputs
model = Model(3).double().cuda()


In [72]:
#Create the loss function and the optimizer
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)


In [73]:
#Define train function
def train(model, train_loader, criterion, optimizer, epochs):
    model.train()
    for epoch in tqdm(range(epochs)):
        for batch_idx, (data, target) in tqdm(enumerate(train_loader)):
            data, target = data.cuda(), target.cuda()
            optimizer.zero_grad()
            output = model(data)
            loss = criterion(output, target)
            loss.backward()
            optimizer.step()
            if batch_idx % 100 == 0:
                print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                    epoch, batch_idx * len(data), len(train_loader.dataset),
                    100. * batch_idx / len(train_loader), loss.item()))


In [74]:
#Define test function
def test(model, test_loader, criterion):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.cuda(), target.cuda()
            output = model(data)
            test_loss += criterion(output, target).item()
            pred = output.argmax(dim=1, keepdim=True)
            correct += pred.eq(target.view_as(pred)).sum().item()
    test_loss /= len(test_loader.dataset)
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


In [75]:
#Train the model
train(model, train_loader, criterion, optimizer, 25)


Train Epoch: 0 [0/22993 (0%)]	Loss: 0.705724
Train Epoch: 0 [6400/22993 (28%)]	Loss: 0.592912
Train Epoch: 0 [12800/22993 (56%)]	Loss: 0.655360
Train Epoch: 0 [19200/22993 (83%)]	Loss: 0.565280
Train Epoch: 1 [0/22993 (0%)]	Loss: 0.499678
Train Epoch: 1 [6400/22993 (28%)]	Loss: 0.516660
Train Epoch: 1 [12800/22993 (56%)]	Loss: 0.591309
Train Epoch: 1 [19200/22993 (83%)]	Loss: 0.468335
Train Epoch: 2 [0/22993 (0%)]	Loss: 0.454736
Train Epoch: 2 [6400/22993 (28%)]	Loss: 0.398781
Train Epoch: 2 [12800/22993 (56%)]	Loss: 0.576110
Train Epoch: 2 [19200/22993 (83%)]	Loss: 0.517679
Train Epoch: 3 [0/22993 (0%)]	Loss: 0.411307
Train Epoch: 3 [6400/22993 (28%)]	Loss: 0.500073
Train Epoch: 3 [12800/22993 (56%)]	Loss: 0.542112
Train Epoch: 3 [19200/22993 (83%)]	Loss: 0.448598
Train Epoch: 4 [0/22993 (0%)]	Loss: 0.508184
Train Epoch: 4 [6400/22993 (28%)]	Loss: 0.435625
Train Epoch: 4 [12800/22993 (56%)]	Loss: 0.425460
Train Epoch: 4 [19200/22993 (83%)]	Loss: 0.455429
Train Epoch: 5 [0/22993 (0%)]	

In [76]:
#Test the model
test(model, test_loader, criterion)



Test set: Average loss: 0.0032, Accuracy: 2379/2555 (93%)



In [ ]:
#Use the model to predict the rain
model.eval()
with torch.no_grad():
    for data, target in test_loader:
        data, target = data.cuda(), target.cuda()
        output = model(data)
        pred = output.argmax(dim=1, keepdim=True)
        print(pred)


In [79]:
#Save the model
torch.save(model.state_dict(), "model.pth")


In [ ]:
#Load the model
model.load_state_dict(torch.load("model.pth"))


---
### 4.2 Descripción de los métoddos y su implementación en código en forma de narrativa
---
Las librerias principales utilizadas fueron en primer lugar Pandas para la manipulación de los datos,Numpy para las transformaciones matematicas así como la construcción de los objetos matematicos para poder , luego Pytorch para la creación del modelo y la implementación de la red neuronal,
Para asegurarnos de que PyTorch haya sido importado correctamente creamos un tensor con 5 números aleatorios dentro de una distribución normal y lo cargamos a CUDA, una arquitectura unificada que permite usar las unidades de procesamiento de gráficos para procesamiento de propósitos generales.
Enseguida creamos un dataframe de pandas al que le cargamos la lectura de la base de datos en CSV que contiene los datos del clima en Seatle de 1948 – 2017, y lo mostramos.
Dentro del dataframe eliminamos las filas que contienen datos “NaN” y reiniciamos los índices evitando que se añadan como una columna más al dataframe.
Después, borramos los datos con el header ‘DATE’ en el eje vertical dentro del mismo dataframe.
Convertimos todos los datos del dataframe en numeros decimales flotantes, para después convertir todo el dataframe en un arreglo numérico de numpy.
Creamos una clase llamada “Dataset” de tipo dataset. Definimos su método de inicialización con el parámetro data. El atributo data será igual al parámetro data ingresado. Guardamos el tamaño del eje horizontal de la data ingresada en el atributo “len”. Creamos un tensor de un arreglo de numpy con la data ingresada hasta la columna 3 (usamos columnas 0, 1 y 2) y lo guardamos en el atributo “x_data”. Creamos otro tensor de un arreglo de numpy con la data ingresada en la columna 3 y lo guardamos en el atributo “y_data”, mismo que convertimos a un tensor de tipo long (para datos enteros de hasta 64 bits, negativos y positivos). Definimos también un método llamado “getitem” al cual ingresa un índice y devuelve el valor de x_data y y_data en el índice especificado. Por último, creamos un método “len” que devuelve el valor de “len” obtenido en el método de inicialización.
Creamos dos datasets, uno de training que utilice la data del arreglo numérico de numpy hasta el 90% de su tamaño, redondeando a un número entero, y uno de testing que utilice el 10% restante redondeado a un número entero.
Utilizando los datasets de training y testing, creamos los dataloaders respectivos con un batch_size de 64 y abilitamos el shuffle para revolver los datos.
Creamos una clase llamada “Model” del tipo nn.Module. Definimos su método de inicialización con el parámetro n_inputs y heredamos los atributos de la clase superior. Aplicamos una transformación lineal utilizando el valor de n_inputs como tamaño de las muestras de entrada y especificamos 3 muestras de salida y guardamos el valor en “hidden1”. Replicamos el proceso dos veces más utilizando ahora 3 para el tamaño de muestras de entrada y de salida, guardando los valores en “hidden2” y “hidden3” respectivamente. Aplicamos una transformación lineal más, pero esta vez cambiamos el tamaño de muestras de salida a 2, para crear nuestra capa de salida, y guardamos su valor en “output”. Por último, en este método, asignamos la función de unidad lineal rectificada al atributo “relu” para utilizarla como función de activación. Definimos después otro método llamado “forward” con el parámetro x. En este guardaremos el valor de x resultante del uso de la función de activación en las tres capas ocultas, y de su paso por la capa de salida. Devolvemos x como salida del modelo.
Creamos un objeto de la clase “Model” con un parámetro de n_inputs = 3 que trabaje con números decimales flotantes y lo cargamos a CUDA. Creamos también un objeto que contiene la función de costo, y un objeto que contiene el algoritmo de optimización “Adam”, con los parámetros del modelo y un learning rate de 0.001, como tensor.

# **5. Resultados**

# **6. Discusión**
---
Inicialmente el programa tenia problemas en la parte del entrenamiento y la evaluación creemos que fue por algún algoritmo mal planteado en la red o incluso alguna función mal planteada. Sin embargo, al revisar minuciosamente encontramos que habíamos incluido los labels en los datos de entrada. Procedimos a reacomodar nuestros datos, ahora en forma de datasets para que fuéramos capaces de generar los tensores de data y labels. Como convertimos nuestros datos a flotantes, tuvimos que adaptar el modelo para funcionar con el mismo tipo de dato. Decidimos utilizar un optimizador diferente y separar las funciones e entrenamiento y validación de los ciclos respectivos para probar otro método y mantener el modelo lo más versátil posible entro de nuestro entendimiento.
El modelo fue probado utilizando 10, 20, 25, 30, 50 y 100 épocas, cada una dando resultados diferentes. Los resultados al correrse con 10 épocas mostraban una precisión baja (56%). Decidimos entonces probar lo que consideramos un exceso de épocas, por los datos que procesamos. Con 100 épocas obtuvimos una precisión del 100%, lo que nos llevó a desconfiar del modelo, ya que podría indicar memorización y no aprendizaje. A partir de estos resultados iniciales de underfitting y overfitting decidimos aproximarnos a un valor que indicara aprendizaje, poca pérdida y alta precisión. 25 épocas nos brindó estas características, por lo que lo consideramos la medida correcta para el entrenamiento de nuestro modelo. Cabe destacar, que cada entrenamiento y testeo se llevaron a cabo con modelos limpios para evitar errores y mantener la integridad de los procesos.


# **7. Conclusiones**
---
En el salón de clases sentí que la creación de un modelo era una tarea sencilla. Me di cuenta de que en realidad lo es, siempre y cuando se tenga una fuerte base de conocimiento derivado del análisis del problema que queremos resolver y las metodologías implementadas en el estado del arte, conocimiento con el que yo no contaba. Analizar los errores, buscar soluciones, aprender de las librerías y su documentación es un proceso que lleva su tiempo y requiere de cierta disposición al replanteamiento de las ideas y conceptos previos para construir nuevos. Sé bien que de este punto en adelante me encontraré con estos conflictos una y otra vez, por lo que entrenar mi plasticidad de adaptabilidad de ideas será un reto diario, no obstante, me llevará a poder crear buenos cimientos para la proposición y construcción de proyectos con soluciones viables por métodos de deep learning.

Al finalizar el modelo de predicción de la lluvia no se cumplieron los objetivos tal cual se esperaba inicialmente, sin embargo, con algo de esfuerzo se obtuvo lo esperado, fue un reto trabajar con este tipo de ANN en múltiples capas y la red neuronal, además para aterrizar lo aprendido y aprender a usar git hub, me pareció una experiencia bien.

En particular Este trabajo demoestró una complejidad y un desafío en la implementación de redes neuronales en la plataforma pytorch y si bien los resultados no fueron satisfactorios fue una buena experiencia para aprender a utilizar esta plataforma y sus herramientas.

# **8. Referencias**

[1] Pb, V. (2020, February 18). Perceptron. Kaggle. Retrieved June 1, 2022, from https://www.kaggle.com/code/prashfio/perceptron/notebook